### ISSUE TITLE
Can't load Model with SparseTensor inputs if it calls a Sequential model internally

### ISSUE THREAD

**Issue by danielegrattarola** [tensorflow/tensorflow#45652]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7

**Describe the current behavior**

When using model-sublcassing to create a `Model` that: 

1. Has a SparseTensor input
2. Uses a Sequential model as internal layer

the model cannot be loaded using `tf.keras.models.load_model`.

The issue is twofold, and I believe it is related to how the signature of `call` is computed upon saving/loading. 

The first problem is that `SparseTensorSpec` does not have a `name` attribute and thus the call to

https://github.com/tensorflow/tensorflow/blob/dd3499e26c7dceaad5333ea0762903de242150b5/tensorflow/python/keras/saving/saved_model/load.py#L812

fails when it has to deal with sparse tensors. 

The second problem is that the signature of the model cannot be properly restored if it uses a Sequential model internally. This, I can't explain. 
Whatever the cause, removing the Sequential model causes the `load_model` method to take this path instead: 

https://github.com/tensorflow/tensorflow/blob/dd3499e26c7dceaad5333ea0762903de242150b5/tensorflow/python/keras/saving/saved_model/load.py#L814

thus avoiding the crash. 

The title of the issue is a bit weird, I didn't know how to describe the problem concisely :D

**Describe the expected behavior**

A Model's signature should be independent of the sub-modules used in its call function and the saving/loading API should support SparseTensors. 

I am happy to contribute code to solve this issue, but I would need some guidance on how to best proceed. 

**Standalone code to reproduce the issue**
```py
import tensorflow as tf
from tensorflow.keras import Model, Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import save_model, load_model


class Works(Model):
    def build(self, input_shape):
        self.fc = Dense(32)

    def call(self, inputs):
        a, b = inputs
        b = self.fc(b)
        return tf.sparse.sparse_dense_matmul(a, b)


class Crashes(Model):
    def build(self, input_shape):
        self.fc = Sequential([Dense(32)])  # <<<< THIS IS THE ONLY DIFFERENCE

    def call(self, inputs):
        a, b = inputs
        b = self.fc(b)
        return tf.sparse.sparse_dense_matmul(a, b)

# Inputs
a = tf.sparse.from_dense(tf.ones((10, 10)))
b = tf.ones((10, 10))

# This works OK, no Sequential model
works = Works()
works([a, b])
save_model(works, 'works')
load_model('works')

# This crashes, it uses a Sequential model
crashes = Crashes()
crashes([a, b])
save_model(crashes, 'crashes')
load_model('crashes')  # <<<< FAILS

```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

```py
Traceback (most recent call last):
  File "~/dev/test.py", line 38, in <module>
    load_model('crashes')
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/save.py", line 187, in load_model
    return saved_model_load.load(filepath, compile, options)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 121, in load
    path, options=options, loader_cls=KerasObjectLoader)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/saved_model/load.py", line 633, in load_internal
    ckpt_options)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 194, in __init__
    super(KerasObjectLoader, self).__init__(*args, **kwargs)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/saved_model/load.py", line 130, in __init__
    self._load_all()
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 221, in _load_all
    self._finalize_objects()
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 526, in _finalize_objects
    _finalize_saved_model_layers(layers_revived_from_saved_model)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 706, in _finalize_saved_model_layers
    inputs = infer_inputs_from_restored_call_function(call_fn)
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 985, in infer_inputs_from_restored_call_function
    spec = nest.map_structure(common_spec, spec, spec2)
  File ~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/util/nest.py", line 635, in map_structure
    structure[0], [func(*x) for x in entries],
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/util/nest.py", line 635, in <listcomp>
    structure[0], [func(*x) for x in entries],
  File "~/dev/tf2/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/load.py", line 981, in common_spec
    x.dtype, x.name)
AttributeError: 'SparseTensorSpec' object has no attribute 'name'

```

**Comment by bhack** [comment]
It was resolved in TF 2.4 please install  `pip install tensorflow==2.4.0-rc3` and run your example again.

**Comment by Saduf2019** [comment]
@daniel-s-ingram 
I ran the code on nightly and tf 2.4 and do not face any issues, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/147fc3c9db3a06bcf3b06b4daa42a347/untitled482.ipynb)
